### PR TITLE
Player knockback

### LINKED ISSUE #391: Inaccurate 1.8 Knockback
I'd like to have the 1.8 knockback in 1.12.2. There are some changes in de velocity of the player being hit. Y-Axis and combos are not like in 1.8

Add a feature ode module to enable that

### LINKED ISSUE #469: Disabling netherite armor's knockback resistance.
As the title says, being able to disable netherite armor's knockback resistance.

Why? Because Netherite's knockback resistance breaks PvP in a way that combos are not too good.

Exemple:

```
disable-knockback-resistance:
  # Disables netherite armor's knockback resistance.
  enabled: true
```

### PR DESCRIPTION
**Pull request by MWHunter**
Attempt to fix #391 and fixes #469 

There's something wrong with knockback still.  The formula is correct but knockback isn't exactly 1.8... probably another issue with gravity changing.  Also, there should be a config to re-enable knockback resistance from netherite and other custom items.

### CODE REVIEW

**Comment by MWHunter** [comment]
I found the bugs causing incorrect 1.8 knockback.  It was applying the vertical knockback limit too soon and not applying friction, which I assumed friction to have already been applied which was incorrect and since knockback is in too different classes, I mixed the order up.  I've tested this on production without finding issues so far, although it probably does need more testing on all versions rather than just 1.16.

I checked the knockback by listening to the knockback packet on paper 1.8 and OldCombatMechanics 1.16, and they are identical!  It isn't an issue with gravity again like I originally thought.  My testing was me standing and hitting a player while walking and sprinting on 1.8 and OCM 1.16 with 1.16 clients, also showing identical results.  I did not test accuracy while moving due to the complexity of it.  I should probably test 1.8 clients, but it should have the same result.

If you listen for the velocity packets on a hit directly on the Z axis, you should see that (units in 1/8000 of block):

On 1.8 walking:
> y  = 2886
> z = 3200

On 1.8 sprinting
> y = 3686
> z = 7200

On OCM 1.16 walking:
> y  = 2886
> z = 3200

On OCM 1.16 sprinting
> y = 3686
> z = 7200

I still need a config option for netherite, forgot that existed at the time that I marked this ready for review.

**Comment by IsXGeo** [comment]
Apologies for asking, when will this be added to the main plugin? My players are asking for it. Thanks

**Comment by DefinePvP** [comment]
> Apologies for asking, when will this be added to the main plugin? My players are asking for it. Thanks

Not sure.  I still haven’t finished disabling netherite knockback since it is client sided.  I’ve tried changing item and player attributes but I must be doing stuff wrong.

I’ve been running my fork in production and everything is fine.

**Comment by IsXGeo** [comment]
I see, would it be possible for the netherite knockback to be toggleable, I would like to keep that feature on my server.

**Comment by MWHunter** [comment]
I found a way to disable netherite knockback with this code

>        Bukkit.getScheduler().runTaskTimer(plugin, () -> {
>
>           for (Player player : Bukkit.getOnlinePlayers()) {
>                new ArrayList<>(player.getAttribute(Attribute.GENERIC_KNOCKBACK_RESISTANCE).getModifiers()).
>                        forEach(modifier -> player.getAttribute(Attribute.GENERIC_KNOCKBACK_RESISTANCE).removeModifier(modifier));
>            }
>        }, 1, 1);

I'll try to figure out the "proper" way to do it but it is possible

**Comment by MWHunter** [comment]
There, I found a way to disable netherite knockback that was cleaner.  It needs some more testing to see if knockback is consistent between versions, which it should since I don't rely on any server code.

This PR is ready to merge after I do some more testing on versions other than 1.16

**Comment by MWHunter** [comment]
Tested vanilla knockback on 1.8 base paper, and on 1.9-1.16 paper oldcombatmechanics with this fork.  Works on all versions OCM supports.  Currently testing in production on airplane 1.16.5 with 100+ players and nothing bad has happened yet.  Assume everything works unless I tell you otherwise.

Ready to merge

File: src/main/java/kernitus/plugin/OldCombatMechanics/module/ModulePlayerKnockback.java [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+import org.bukkit.event.entity.EntityDamageByEntityEvent;
+import org.bukkit.event.entity.EntityDamageEvent;
+import org.bukkit.event.player.PlayerVelocityEvent;
+import org.bukkit.util.Vector;
+
+import java.util.HashMap;
+
+public class ModulePlayerKnockback extends Module {
+    /**
+     * The knockbacck formula is changed in 1.9, especially with sprinting
```
Comment (Aberdeener) [inline src/main/java/kernitus/plugin/OldCombatMechanics/module/ModulePlayerKnockback.java:21]: knockback*

**Comment by MWHunter** [comment]
Two things that could be changed:
- Disabling netherite for all damage sources, which would make thorns/fishing hooks do the right knockback.  Should be worth the small performance impact.
- On 1.9-1.15 this branch tries to remove an attribute that doesn't exist.  Due to bukkit's implementation, this doesn't cause an error (a new arraylist is always created) but there is a slight performance benefit to check the version, and badly coded 1.8-1.15 forks could change this behavior and break the plugin.

Everything works but it could be better.

**Pull request merged**